[Phase-2 L1] Incorporate object threshold scalings, ID and isolation via dictionaries into the P2GT menu

artlbv opened the pull request:
#### PR description:

The idea of this PR is to allow for a more convenient organisation and book-keeping of the P2GT emulator menu configurations. Previously threshold changes due to object changes (so-called offline-online scalings) or ID/Iso WP changes had to be changed manually in every seed. With this PR dictionaries are introduced for common ID and Isolation Working Points as well as reading of the scalings (a MenuTools output) is implemented.

This addresses #46470 and was agreed to with the P2GT developer team @haarigerharald and @qvyz as presented [here in slide 42](https://indico.cern.ch/event/1469447/contributions/6186551/attachments/2951766/5190339/P2L1Menu%20Spring24%20MC%20Report%20October%202024.pdf).

FYI @VourMa @rovere @slaurila @robertjward @gpetruc 

#### PR validation:

Tested to work with the standard recipe we (and HLT) are using in `1420pre1`:
```
cmsDriver.py l1nanoPhase2 -s L1,L1TrackTrigger,L1P2GT,USER:PhysicsTools/L1Nano/l1tPh2Nano_cff.l1tPh2NanoTask --customise PhysicsTools/L1Nano/l1tPh2Nano_cff.addFullPh2L1Nano \                 
--conditions auto:phase2_realistic_T33 \
--geometry Extended2026D110 \
--era Phase2C17I13M9 \
--eventcontent NANOAOD \
--datatier GEN-SIM-DIGI-RAW-MINIAOD \
--customise SLHCUpgradeSimulations/Configuration/aging.customise_aging_1000,Configuration/DataProcessing/Utils.addMonitoring,L1Trigger/Configuration/customisePhase2TTOn110.customisePhase2TTOn110 \
--filein /store/mc/Phase2Spring24DIGIRECOMiniAOD/TT_TuneCP5_14TeV-powheg-pythia8/GEN-SIM-DIGI-RAW-MINIAOD/PU200_AllTP_140X_mcRun4_realistic_v4-v1/2560000/11d1f6f0-5f03-421e-90c7-b5815197fc85.root \
--fileout file:output_Phase2_L1T.root \
--python_filename rerunL1_cfg.py \
--inputCommands="keep *, drop l1tPFJets_*_*_*, drop l1tTrackerMuons_l1tTkMuonsGmt*_*_HLT" \
--mc \
-n 1000 --nThreads 4
```

A simple check was comparing the `edmConfigDump` outputs pre and post PR as there the actual thresholds, ID and Isolation values are expanded into plaintext. The results seem reasonable.
Note that also the IDs used for TkMuons have changed since since the AR24 there are now 4 WPs (compared to 1 before).

We will check the agreement with the MenuTools once we get more stats processed (on small stats the agreement is rather good), but I would like to start the PR and the reviewing process, eventually soliciting feedback from our SW experts on these changes. 

Note that the code function names in the `menuConstants.py` could be cleaned up etc, but the mechanism should the final one.

cmsbuild commented:
cms-bot internal usage<!-- bot cache: {"commits":{"0b37b83074490649c4e887173cc40a88317d3414":{"files":["L1Trigger/Phase2L1GT/python/l1tGTMenu_hadr_crossLepSeeds_cff.py","L1Trigger/Phase2L1GT/python/l1tGTMenu_lepSeeds_cff.py"],"squashed":false,"time":1729603131},"1bc87802cf557c68575da0e52ab5168d6a99dce0":{"files":["L1Trigger/Phase2L1GT/python/l1tGTMenu_BTagSeeds_cff.py"],"squashed":false,"time":1729608615},"2a586eba7ebbe95785abf778aae117ff2181dee3":{"files":["L1Trigger/Phase2L1GT/python/l1tGTMenu_lepSeeds_cff.py","L1Trigger/Phase2L1GT/python/menuConstants.py"],"squashed":false,"time":1729594608},"325d4e7f9e6cfebca90f9b065b779143df4cb2ce":{"files":["L1Trigger/Phase2L1GT/python/l1tGTMenu_BTagSeeds_cff.py"],"squashed":false,"time":1729604558},"4e8522d99c077f144ede307037e1e22e014ecabf":{"files":["L1Trigger/Phase2L1GT/python/l1tGTMenu_lepSeeds_cff.py","L1Trigger/Phase2L1GT/python/menuConstants.py"],"squashed":false,"time":1729591805},"5e957410515a1bc6a05acb9d7379ec076d0bed07":{"files":["L1Trigger/Phase2L1GT/python/l1tGTMenu_hadr_metSeeds_cff.py"],"squashed":false,"time":1729609103},"6ee092d8795100a4526b2ebd6b484abc794f81a0":{"files":["L1Trigger/Phase2L1GT/python/l1tGTMenu_lepSeeds_cff.py"],"squashed":false,"time":1728573493},"a83f4bd4159a8be9dda49d3080fd21c3f7cca745":{"files":["L1Trigger/Phase2L1GT/python/l1tGTMenu_crossLepSeeds_cff.py"],"squashed":false,"time":1729601819},"ad00aa35a1c2446942be89efe1b19602daa11537":{"files":["L1Trigger/Phase2L1GT/python/l1tGTMenu_lepSeeds_cff.py","L1Trigger/Phase2L1GT/python/menuConstants.py"],"squashed":false,"time":1729592514},"b41ce94e10a478cc94aa341f3f6b2fc9d95bdaac":{"files":["L1Trigger/Phase2L1GT/python/l1tGTMenu_lepSeeds_cff.py","L1Trigger/Phase2L1GT/python/menuConstants.py"],"squashed":false,"time":1729593541},"edcb741535a9b83ec24a04260b9fd5aeca1a1c2b":{"files":["L1Trigger/Phase2L1GT/python/menuConstants.py"],"squashed":false,"time":1729605227},"f32e2c65fa2ecd2573ea6db2386bec30b82c2842":{"files":["L1Trigger/Phase2L1GT/python/l1tGTMenu_hadr_crossLepSeeds_cff.py","L1Trigger/Phase2L1GT/python/l1tGTMenu_hadr_metSeeds_cff.py","L1Trigger/Phase2L1GT/python/menuConstants.py"],"squashed":false,"time":1729604373},"f45db124d18326ac2db3fc53a345a094d3501867":{"files":["L1Trigger/Phase2L1GT/python/l1tGTMenu_lepSeeds_cff.py","L1Trigger/Phase2L1GT/python/menuConstants.py","L1Trigger/Phase2L1GT/python/scalings_v44.json"],"squashed":false,"time":1729587473}},"emoji":{},"last_seen_sha":"5e957410515a1bc6a05acb9d7379ec076d0bed07","signatures":{}} -->

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-46489/42345

cmsbuild commented:
A new Pull Request was created by  @artlbv for master.

It involves the following packages:

- L1Trigger/Phase2L1GT (**upgrade, l1**)


@Moanwar, @aloeliger, @cmsbuild, @epalencia, @srimanob, @subirsarkar can you please review it and eventually sign? Thanks.
@Martin-Grunewald, @missirol, @mmusich this is something you requested to watch as well.
@antoniovilela, @mandrenguyen, @rappoccio, @sextonkennedy you are the release manager for this.

cms-bot commands are listed <a href="http://cms-sw.github.io/cms-bot-cmssw-cmds.html">here</a>

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-46489/42349

cmsbuild commented:
Pull request #46489 was updated. @Moanwar, @aloeliger, @cmsbuild, @epalencia, @srimanob, @subirsarkar can you please check and sign again.

mmusich commented:
Does this incorporate the actual fix for

> For example, pDoublePuppiTau52_52 in our measurements has ~3 times the rate of pSingleTkEle36

as discussed in https://github.com/cms-sw/cmssw/issues/46470?

mmusich commented:
test parameters
   * enable hlt_p2_timing

mmusich commented:
@cmsbuild, please test

mmusich commented:
@cmsbuild, please abort

mmusich commented:
please test

artlbv commented:
> Does this incorporate the actual fix for
> 
> > For example, pDoublePuppiTau52_52 in our measurements has ~3 times the rate of pSingleTkEle36
> 
> as discussed in #46470?

Yes. At least the P2GT thresholds are now as in the MenuTools, and for the MenuTools we see similar rates to what we had in the AR24, see our recent presentation at the L1 DPG ([linked above](https://indico.cern.ch/event/1469447/#28-phase-2-sample-validation-f), slide 38). We will eventually also rerun the full MinBias but the good match of the MenuTools and P2GT emulator on 150 TTbar events makes me positive about a good agreement.

cmsbuild commented:
+1

**Size**: This PR adds an extra 20KB to repository
**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-d2424a/42363/summary.html
**COMMIT**: e2e4b6ceb091cd04e1f34af547d2b0173b148458
**CMSSW**: CMSSW_14_2_X_2024-10-23-1100/el8_amd64_gcc12
**Additional Tests**: HLT_P2_TIMING
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week0/cms-sw/cmssw/46489/42363/install.sh` to create a dev area with all the needed externals and cmssw changes.


* **HLT P2 Timing**: [chart](https://cmssdt.cern.ch/circles/web/piechart.php?data_name=hlt-p2-timing&resource=time_thread&filter=CMSSW_14_2_X_2024-10-23-1100&dataset=CMSSW_14_2_X_2024-10-23-1100-cmssw46489/el8_amd64_gcc12/33/Phase2Timing_resources)

## Comparison Summary



**Summary**:
 - You potentially [added 3 lines](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_14_2_X_2024-10-23-1100+d2424a/65258/validateJR/logRootQA-events.log) to the logs
 - [Reco comparison results: 123 differences found in the comparisons](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_14_2_X_2024-10-23-1100+d2424a/65258/validateJR.html)
 - DQMHistoTests: Total files compared: 46
 - DQMHistoTests: Total histograms compared: 3566331
 - DQMHistoTests: [Total failures: 415](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_14_2_X_2024-10-23-1100+d2424a/65258/validateJR.html)
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 3565896
 - DQMHistoTests: Total skipped: 20
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 45 files compared)
 - Checked 201 log files, 171 edm output root files, 46 DQM output files
 - [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_14_2_X_2024-10-23-1100+d2424a/65258/triggerResults): found differences in 4 / 44 workflows

rovere commented:
The performance aligns with the measurements I did privately using only the fix for the DiTau52 path: +10% speed-up and a visible reduction of events triggering HLT Tau paths.

mmusich commented:
Indeed, putting here some details for the record as the PR test results, alas, are not forever.

### Timing

| `CMSSW_14_2_X_2024-10-23-1100` | `CMSSW_14_2_X_2024-10-23-1100` + this PR |
| ------------- | ------------- |
|  ![Screenshot from 2024-10-24 08-44-52](https://github.com/user-attachments/assets/a9e331ee-bd55-426f-bd8a-9e8d8f16448d) | ![Screenshot from 2024-10-24 08-44-42](https://github.com/user-attachments/assets/1aa646cd-9ee0-4e22-ad27-62008bf54ac7) |

baseline: 5304.3 ms/ev | this PR: 4799.6 ms/ev | speedup gain **-9.51%**

### Trigger Results

> [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_14_2_X_2024-10-23-1100+d2424a/65258/triggerResults): found differences in 4 / 44 workflows

e.g. for workflow [29634.0](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_14_2_X_2024-10-23-1100+d2424a/65258/triggerResults/29634.0_TTbar_14TeV+2026D110)

```
Found 10 matching events, out of which 6 have different HLT results

      Events    Accepted      Gained        Lost       Other  Trigger
          10           6           -          -1           -  pDoublePuppiJet112_112
          10           7           -          -2           -  pDoublePuppiTau52_52
          10           4           -          -2           -  pNNPuppiTauPuppiMet_55_190
          10           4          +1           -           -  pPuppiHT450
          10           5           -          -2           -  pSinglePuppiJet230
          10           0           -           -          ~2  HLT_AK4PFPuppiJet520
          10           0           -           -          ~1  HLT_PFPuppiHT1070
          10           1           -           -          ~1  HLT_DoublePFPuppiJets128_DoublePFPuppiBTagDeepCSV_2p4
          10           2           -           -          ~1  HLT_DoublePFPuppiJets128_DoublePFPuppiBTagDeepFlavour_2p4
          10           1           -           -          ~2  HLT_DoubleMediumChargedIsoPFTauHPS40_eta2p1
          10           3           -           -          ~2  HLT_DoubleMediumDeepTauPFTauHPS35_eta2p1
```

mmusich commented:
type bug-fix

aloeliger commented on L1Trigger/Phase2L1GT/python/menuConstants.py:
```diff
@@ -0,0 +1,103 @@
+import json
+
+# print(os.path.exists("./scalings_v44.json"))
+scalings_fname = environ["CMSSW_BASE"] + "/src/L1Trigger/Phase2L1GT/python/scalings_v44.json"
+with open(scalings_fname, 'r') as file:
+    scalings = json.load(file)
```
I don't think you are necessarily guaranteed this is always going to exist (with a partial checkout of CMSSW for example where you are not working with these directories directly). 

The JSON itself should be committed as a data file as it is not directly code.

I'm not sure if there is a FileInPath solution for inside a python configuration, but maybe one of the core people know.

artlbv replied:
Should it be the `RELEASE_BASE` then instead? 

On the separation to `cms-data`: I think that given these constants are tied to the L1 menu, it would make sense to keep them together imo. Otherwise it might be hard to disentangle if changes are due to the menu config or constant changes.

mmusich replied:
> I'm not sure if there is a FileInPath solution for inside a python configuration, but maybe one of the core people know.

there is `pfnInPath`.

 > Otherwise it might be hard to disentangle if changes are due to the menu config or constant changes.
 
I am not sure to see the issue. Can you elaborate? 
Having said that, the main draw-back of storing the configuration in `cms-data` is that a configuration change will require a full release as opposed to a patch-release. Having said that it's not clear to me what's the long term strategy for storing the phase-2 L1T menu, so it's hard to say if that is going to be an issue.

artlbv replied:
@mmusich do you mean using this for example?
https://github.com/cms-sw/cmssw/blob/902185ebb53bc03dd80c6bfd5cb50a5667d572e3/Alignment/OfflineValidation/python/TkAlAllInOneTool/findAndChange.py#L55

artlbv replied:
Moved the constants to a separate python file and thus can load it via the standard import.

aloeliger commented on L1Trigger/Phase2L1GT/python/menuConstants.py:
```diff
@@ -0,0 +1,103 @@
+    },
+    "L1tkElectron":{
+        "Iso": {
+            # "qual": {
+            #     "barrel": 0b0010,
+            #     "endcap": 0b0010,
+            # },
+            "iso": {
+                "barrel": 0.13,
+                "endcap": 0.28,
+            }
+        },
+        "NoIso": {
+            "qual": {
+                "barrel": 0b0010,
+                "endcap": 0b0010,
+            },
+        },
+        "NoIsoLowPt": {
+            "qual": {
+                "barrel": 0b0010,
+                "endcap": 0b0000,
+            },
+        }
+    },
+    "L1gmtTkMuon":{
+        "VLoose": {
+            "qual": 0b0001,
+        },
+        "Loose": {
+            "qual": 0b0010,
+        },
+        "Medium": {
+            "qual": 0b0100,
+        },
+        "Tight": {
+            "qual": 0b1000,
+        },
+    }
+}
```
If we are already storing menu constants in JSON, could this also be added to the JSON?

artlbv replied:
Yes, I agree. though it would be a separate JSON, as the one with scalings is generated from the MenuTools.

artlbv replied:
Moved this to a separate python file

gpetruc commented:
Hi,

For the long term strategy of storing the L1T Data menu, it's to be defined
but it's not going to be files in the release.
However, for MCs where just a fixed menu is included, a file in the release
is probably good enough (it has been so with HLT).

To be better defined later, e.g. if the Phase-2 L1T has an HLT-like
structure then we might as well use the same tech to store and version data
menus for both.

Giovanni

Il Lun 28 Ott 2024, 15:33 Marco Musich ***@***.***> ha
scritto:

> ***@***.**** commented on this pull request.
> ------------------------------
>
> In L1Trigger/Phase2L1GT/python/menuConstants.py
> <https://github.com/cms-sw/cmssw/pull/46489#discussion_r1819182491>:
>
> > +scalings_fname = environ["CMSSW_BASE"] + "/src/L1Trigger/Phase2L1GT/python/scalings_v44.json"
> +with open(scalings_fname, 'r') as file:
> +    scalings = json.load(file)
>
> I'm not sure if there is a FileInPath solution for inside a python
> configuration, but maybe one of the core people know.
>
> there is pfnInPath.
>
> Otherwise it might be hard to disentangle if changes are due to the menu
> config or constant changes.
>
> I am not sure to see the issue. Can you elaborate?
> Having said that, the main draw-back of storing the configuration in
> cms-data is that a configuration change will require a full release as
> opposed to a patch-release. Having said that it's not clear to me what's
> the long term strategy for storing the phase-2 L1T menu, so it's hard to
> say if that is going to be an issue.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/cms-sw/cmssw/pull/46489#discussion_r1819182491>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABDNWR7LCBTVNLSFZLZPDEDZ5Y4MDAVCNFSM6AAAAABQOTRMAOVHI2DSMVQWIX3LMV43YUDVNRWFEZLROVSXG5CSMV3GSZLXHMZDGOJZGI4DGMZTGQ>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

aloeliger commented:
@artlbv what is that status of this PR?

artlbv commented on L1Trigger/Phase2L1GT/python/scalings_v44.json:
1. do you mean we should just save the JSON with `.py` as extension?
2. I agree and was planning to remap the object names in the JSON – will push this for this PR

HaarigerHarald replied:
1. Well you need a variable to assign the constant but other than that yes. Now, I understand that you had some automation with the `json` export in mind, but do you see the scales changing that often?
2. Great!

artlbv commented on L1Trigger/Phase2L1GT/python/menuConstants.py:
```diff
@@ -0,0 +1,103 @@
+    else:
+        return max(0, new_thr)
+
+def getObjectThrs(thr, obj, id):
```
this is too long and redundant imo (what if the off. threshold is mistyped in one of the regions?).
a compromise could be to list the regions as another argument, i.e.:
```
regionsMinPt  = getObjectThrs(17, "L1nnPuppiTau", "default", ["barrel","endcap"])
```

how about this?

HaarigerHarald replied:
Looks good to me as well, since one can at least guess that this might return a `cms.vdouble`.

artlbv replied:
I decided to not use explicit regions in each function and instead introduce a global region dict and cross-check against it all the scalings, IDs, iso etc that are picked up –> for the sake of clarity and compactness.

In principle, with the new/restructured  menu writing with baseline objects we could write out the regions explicitly. OTOH as I wrote in the general discussion I think we should just introduce the scalings into the P2GT object and thus simplify the config even further. We should discuss this further, but for this PR imo the state as is should be acceptable.

artlbv commented:
> @artlbv what is that status of this PR?

there were some more pending checks before finishing this PR (+feedback from the P2GT that just came): [see this JIRA ticket](https://its.cern.ch/jira/browse/CMSLITDPG-1326?focusedId=6609464&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-6609464)
– check discrepancies with MenuTools when both use TkPhotons (no standalone EG)
– decide what to do with the dynamic ID of the low pt Muons

I'm not sure how urgent these checks are wrt to merging this PR into master tbh.
@rovere @VourMa do you prefer the PR to go in as is or rather to have the above discrepancies sorted out before?

HaarigerHarald commented:
> > @artlbv what is that status of this PR?
> 
> there were some more pending checks before finishing this PR (+feedback from the P2GT that just came): [see this JIRA ticket](https://its.cern.ch/jira/browse/CMSLITDPG-1326?focusedId=6609464&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-6609464) – check discrepancies with MenuTools when both use TkPhotons (no standalone EG) – decide what to do with the dynamic ID of the low pt Muons
> 
> I'm not sure how urgent these checks are wrt to merging this PR into master tbh. @rovere @VourMa do you prefer the PR to go in as is or rather to have the above discrepancies sorted out before?

The muon quality was already updated as part of https://github.com/cms-sw/cmssw/pull/45606. I ran over 10000 ttbar pre and post change to ensure that both cuts are equivalent.

VourMa commented:
> @rovere @VourMa do you prefer the PR to go in as is or rather to have the above discrepancies sorted out before?

From my side, I think it's better to finalize this PR, as it takes care of the most important issues we have been seeing. Then, we can follow up with the remaining fixes in a separate PR, especially because I understand that, for these extra fixes, there may not be an immediate solution.

cmsbuild commented:
Milestone for this pull request has been moved to CMSSW_15_0_X. Please open a backport if it should also go in to CMSSW_14_2_X.

mmusich commented:
> From my side, I think it's better to finalize this PR, as it takes care of the most important issues we have been seeing. Then, we can follow up with the remaining fixes in a separate PR, especially because I understand that, for these extra fixes, there may not be an immediate solution.

we just got reminded that currently we can't get sound estimates on the timing increase due to new path additions (e.g. in https://github.com/cms-sw/cmssw/pull/46924). Can someone update on the foreseen trajectory for this PR from the L1 side @artlbv @cms-sw/l1-l2 ?

mmusich commented:
@cmsbuild, please test

cmsbuild commented:
+1

**Size**: This PR adds an extra 12KB to repository
**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-d2424a/43505/summary.html
**COMMIT**: e2e4b6ceb091cd04e1f34af547d2b0173b148458
**CMSSW**: CMSSW_15_0_X_2024-12-17-1100/el8_amd64_gcc12
**Additional Tests**: HLT_P2_TIMING
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week0/cms-sw/cmssw/46489/43505/install.sh` to create a dev area with all the needed externals and cmssw changes.


* **HLT P2 Timing**: [chart](https://cmssdt.cern.ch/circles/web/piechart.php?data_name=hlt-p2-timing&resource=time_thread&filter=CMSSW_15_0_X_2024-12-17-1100&dataset=CMSSW_15_0_X_2024-12-17-1100-cmssw46489/el8_amd64_gcc12/64/Phase2Timing_resources)

## Comparison Summary



**Summary**:
 - You potentially [removed 1 lines](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_15_0_X_2024-12-17-1100+d2424a/66180/validateJR/logRootQA-events.log) from the logs
 - [Reco comparison results: 120 differences found in the comparisons](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_15_0_X_2024-12-17-1100+d2424a/66180/validateJR.html)
 - DQMHistoTests: Total files compared: 46
 - DQMHistoTests: Total histograms compared: 3510017
 - DQMHistoTests: [Total failures: 373](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_15_0_X_2024-12-17-1100+d2424a/66180/validateJR.html)
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 3509624
 - DQMHistoTests: Total skipped: 20
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 45 files compared)
 - Checked 202 log files, 172 edm output root files, 46 DQM output files
 - [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_15_0_X_2024-12-17-1100+d2424a/66180/triggerResults): found differences in 4 / 44 workflows

VourMa commented:
Hi @artlbv, I was wondering whether it is still the plan to update this PR these days. The last 15_0_X pre-release deadline is on the 4th of February and, given that 15_0_X is a candidate for our Phase 2 HLT sample production, it is important to have this in by then (apart from it being relevant in all of our timing tests).

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-46489/43327

- Found files with invalid states:
  - L1Trigger/Phase2L1GT/python/scalings_v44.json:
    - Added: f45db124d18326ac2db3fc53a345a094d3501867
    - Deleted: 6897d0cb1d5cae2076065dce548d852b4c70914c

cmsbuild commented:
Pull request #46489 was updated. @Moanwar, @aloeliger, @cmsbuild, @epalencia, @srimanob, @subirsarkar can you please check and sign again.

mmusich commented:
allow @artlbv test rights

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-46489/43334

- Found files with invalid states:
  - L1Trigger/Phase2L1GT/python/l1tGTMenu_constants.py:
    - Added: 75f1e410f42f3591a0ff27c0408b4f7075eaf64e
    - Deleted: ba26ec451f281ff556a3e8179b25f2ae11b5e970
  - L1Trigger/Phase2L1GT/python/menuConstants.py:
    - Added: f45db124d18326ac2db3fc53a345a094d3501867
    - Modified: 4e8522d99c077f144ede307037e1e22e014ecabf, ad00aa35a1c2446942be89efe1b19602daa11537, b41ce94e10a478cc94aa341f3f6b2fc9d95bdaac, 2a586eba7ebbe95785abf778aae117ff2181dee3, f32e2c65fa2ecd2573ea6db2386bec30b82c2842, edcb741535a9b83ec24a04260b9fd5aeca1a1c2b, a06da2cd9d0af476477b79527129dc1d92b5615c, a23de8a11f2a5d4fecc1df1c7cd43ee5a0365c9c
    - Deleted: 75f1e410f42f3591a0ff27c0408b4f7075eaf64e
  - L1Trigger/Phase2L1GT/python/scalings_v44.json:
    - Added: f45db124d18326ac2db3fc53a345a094d3501867
    - Deleted: 6897d0cb1d5cae2076065dce548d852b4c70914c
  - L1Trigger/Phase2L1GT/python/l1tObjectScalings.py:
    - Added: 6897d0cb1d5cae2076065dce548d852b4c70914c
    - Deleted: ba26ec451f281ff556a3e8179b25f2ae11b5e970
  - L1Trigger/Phase2L1GT/python/l1tObjectIDs.py:
    - Added: 6897d0cb1d5cae2076065dce548d852b4c70914c
    - Deleted: ba26ec451f281ff556a3e8179b25f2ae11b5e970

cmsbuild commented:
Pull request #46489 was updated. @Moanwar, @aloeliger, @cmsbuild, @epalencia, @srimanob, @subirsarkar can you please check and sign again.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-46489/43338

- Found files with invalid states:
  - L1Trigger/Phase2L1GT/python/l1tGTMenu_constants.py:
    - Added: 75f1e410f42f3591a0ff27c0408b4f7075eaf64e
    - Deleted: ba26ec451f281ff556a3e8179b25f2ae11b5e970
  - L1Trigger/Phase2L1GT/python/menuConstants.py:
    - Added: f45db124d18326ac2db3fc53a345a094d3501867
    - Modified: 4e8522d99c077f144ede307037e1e22e014ecabf, ad00aa35a1c2446942be89efe1b19602daa11537, b41ce94e10a478cc94aa341f3f6b2fc9d95bdaac, 2a586eba7ebbe95785abf778aae117ff2181dee3, f32e2c65fa2ecd2573ea6db2386bec30b82c2842, edcb741535a9b83ec24a04260b9fd5aeca1a1c2b, a06da2cd9d0af476477b79527129dc1d92b5615c, a23de8a11f2a5d4fecc1df1c7cd43ee5a0365c9c
    - Deleted: 75f1e410f42f3591a0ff27c0408b4f7075eaf64e
  - L1Trigger/Phase2L1GT/python/scalings_v44.json:
    - Added: f45db124d18326ac2db3fc53a345a094d3501867
    - Deleted: 6897d0cb1d5cae2076065dce548d852b4c70914c
  - L1Trigger/Phase2L1GT/python/l1tObjectScalings.py:
    - Added: 6897d0cb1d5cae2076065dce548d852b4c70914c
    - Deleted: ba26ec451f281ff556a3e8179b25f2ae11b5e970
  - L1Trigger/Phase2L1GT/python/l1tObjectIDs.py:
    - Added: 6897d0cb1d5cae2076065dce548d852b4c70914c
    - Deleted: ba26ec451f281ff556a3e8179b25f2ae11b5e970
  - L1Trigger/Phase2L1GT/python/l1GTMenuObjects_cff.py:
    - Added: e22746574cfd2725faa1bb548866b404d4c8af2f
    - Modified: 4359776cc4560347dc2a13ee709303292fc621a0, ba4e4dd3c75fe97dd78448478b07e94905971223
    - Deleted: 793cff1312bebb31653509697e9887b63ed7d3da

cmsbuild commented:
Pull request #46489 was updated. @Moanwar, @aloeliger, @cmsbuild, @epalencia, @srimanob, @subirsarkar can you please check and sign again.

artlbv commented:
Alright, I think I got the PR to a state where I would say that it can be reviewed @aloeliger 

I added some documentation about the new features in the [README](https://github.com/cms-sw/cmssw/blob/e0e95cd0437bf21cff6422e15934e905828177a7/L1Trigger/Phase2L1GT/README.md#object-presets-to-synchronise-with-the-menutools).

While updating the whole menu I realised that it should be possible to further simplify the config if repeated objects would be re-used, hence I completely rewrote the full menu to use this strategy as documented in [this part of the README](https://github.com/cms-sw/cmssw/blob/e0e95cd0437bf21cff6422e15934e905828177a7/L1Trigger/Phase2L1GT/README.md#simplification-for-common-baseline-objects).

Short summary: where before we had to write out the full specs of an object like this
```python
collection1 = cms.PSet(
    tag = cms.InputTag("l1tGTProducer", "GMTTkMuons"),
    regionsAbsEtaLowerBounds=cms.vdouble(0,0.83,1.24),
    regionsMinPt = get_object_thrs(15, "GMTTkMuons","VLoose"),
    qualityFlags = get_object_ids("GMTTkMuons","VLoose"),
    ...
)
```

now this is replaced with:
```python
collection1 = l1tGTtkMuonVLoose.clone(
        regionsMinPt = get_object_thrs(15, "GMTTkMuons","VLoose"),
    ),
```
where `l1tGTtkMuonVLoose ` is defined in `L1Trigger/Phase2L1GT/python/l1tGTMenuObjects_cff.py`

While rewriting the menu I noticed and fixed some minor inconsistencies in the current P2GT emulator menu and thus will need to rerun a larger validation with the MenuTools – that should be doable within next week. I'll launch the tests here nevertheless.

I would personally be very happy if we could avoid writing this odd string `regionsMinPt = get_object_thrs(15, "GMTTkMuons","VLoose"),` for every object, but that would likely require incorporating the scalings into the P2GT collection classes. But at least this cleanup and streamlining of the menu config should be already a much cleaner and more-easy-to-maintain version.

artlbv commented:
please test

cmsbuild commented:
-1

**Failed Tests**: HLTP2Timing 
**Size**: This PR adds an extra 80KB to repository
**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-d2424a/43804/summary.html
**COMMIT**: e0e95cd0437bf21cff6422e15934e905828177a7
**CMSSW**: CMSSW_15_0_X_2025-01-16-2300/el8_amd64_gcc12
**Additional Tests**: HLT_P2_TIMING
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week0/cms-sw/cmssw/46489/43804/install.sh` to create a dev area with all the needed externals and cmssw changes.


## Comparison Summary



**Summary**:
 - You potentially [added 2 lines](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_15_0_X_2025-01-16-2300+d2424a/66473/validateJR/logRootQA-events.log) to the logs
 - [Reco comparison results: 120 differences found in the comparisons](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_15_0_X_2025-01-16-2300+d2424a/66473/validateJR.html)
 - DQMHistoTests: Total files compared: 49
 - DQMHistoTests: Total histograms compared: 3819085
 - DQMHistoTests: [Total failures: 101](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_15_0_X_2025-01-16-2300+d2424a/66473/validateJR.html)
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 3818964
 - DQMHistoTests: Total skipped: 20
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 48 files compared)
 - Checked 214 log files, 184 edm output root files, 49 DQM output files
 - [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_15_0_X_2025-01-16-2300+d2424a/66473/triggerResults): found differences in 4 / 47 workflows

mmusich commented:
> Failed Tests: HLTP2Timing

failures are unrelated, the IB fails as well: [link](https://cmssdt.cern.ch/SDT/jenkins-artifacts/hlt-p2-timing/CMSSW_15_0_X_2025-01-16-2300/el8_amd64_gcc12/run.log). The issue is likely on the testing machine

mmusich commented:
@cmsbuild, please test

cmsbuild commented:
+1

**Size**: This PR adds an extra 80KB to repository
**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-d2424a/43809/summary.html
**COMMIT**: e0e95cd0437bf21cff6422e15934e905828177a7
**CMSSW**: CMSSW_15_0_X_2025-01-16-2300/el8_amd64_gcc12
**Additional Tests**: HLT_P2_TIMING
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week0/cms-sw/cmssw/46489/43809/install.sh` to create a dev area with all the needed externals and cmssw changes.


* **HLT P2 Timing**: [chart](https://cmssdt.cern.ch/circles/web/piechart.php?data_name=hlt-p2-timing&resource=time_thread&filter=CMSSW_15_0_X_2025-01-16-2300&dataset=CMSSW_15_0_X_2025-01-16-2300-cmssw46489/el8_amd64_gcc12/73/Phase2Timing_resources)

## Comparison Summary



**Summary**:
 - No significant changes to the logs found
 - [Reco comparison results: 122 differences found in the comparisons](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_15_0_X_2025-01-16-2300+d2424a/66475/validateJR.html)
 - DQMHistoTests: Total files compared: 49
 - DQMHistoTests: Total histograms compared: 3819085
 - DQMHistoTests: [Total failures: 104](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_15_0_X_2025-01-16-2300+d2424a/66475/validateJR.html)
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 3818961
 - DQMHistoTests: Total skipped: 20
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 48 files compared)
 - Checked 214 log files, 184 edm output root files, 49 DQM output files
 - [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_15_0_X_2025-01-16-2300+d2424a/66475/triggerResults): found differences in 4 / 47 workflows

artlbv commented:
Let me copy the trigger and timing results here for archival reasons as Marco did [above](https://github.com/cms-sw/cmssw/pull/46489#issuecomment-2434439616).

>  [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_15_0_X_2025-01-16-2300+d2424a/66475/triggerResults): found differences in 4 / 47 workflows

    Events    Accepted      Gained        Lost       Other  Trigger
          10           1           -          -1           -  pDoubleEGEle37_24
          10           5           -          -1           -  pDoublePuppiJet112_112
          10           3           -          -1           -  pDoublePuppiTau52_52
          10           2           -          -1           -  pIsoTkEleEGEle22_12
          10           1           -          -1           -  pPuppiTauTkMuon42_18
          10           3           -          -1           -  pSingleEGEle51
          10           1           -          -1           -  pTkMuPuppiJetPuppiMet_3_110_120

I'll compare to a recent PR where the HLT phase2 timing was run : https://github.com/cms-sw/cmssw/pull/47064#issuecomment-2579661596

| [#47064](https://github.com/cms-sw/cmssw/pull/47064) | This PR |
| -------- | -------- |
| ![](https://codimd.web.cern.ch/uploads/upload_806b21742a4f3971a114d087935c4b87.png) | ![](https://codimd.web.cern.ch/uploads/upload_8b3228a405890a3e86feaaf53202bd0c.png)   |

Even though the timing is a bit more than it was in October, there is still about 9% gain from the update of the P2GT menu in this PR.

artlbv commented:
### A final thought from my side on the P2GT menu restructuring: 

I think that it makes sense to move the P2GT menu(s) away from the P2GT producer folder into the L1 Configuration (similar to the HLT test menus).

I made [this branch](https://github.com/cms-sw/cmssw/compare/master...artlbv:p2gt_refact_menus_2024?expand=1) where I moved the current 2023 and the new 2024 menu into `L1Trigger/Configuration`:
https://github.com/artlbv/cmssw/tree/p2gt_refact_menus_2024/L1Trigger/Configuration/python/Phase2GTMenu

* `step1_2023` –> "current" menu as of 2023
* `step1_2024_explicitSeeds` –>  the intermediate one with still explicit definitions of individual objects
* `step1_2024` -> the final version as in this PR currently

I put the final 2024 version as default in the global sequence, though in cmsDriver one can select any other as `-s L1P2GT:step1_2023`.

For me that would be the preferred location, though @aloeliger @HaarigerHarald @qvyz @gpetruc should comment if anything speaks against it.

**If there are no objections, I would close this PR and reopen another one (or should I force push over this branch?).**

FYI @slaurila @VourMa @rovere @RobertJWard

aloeliger commented:
@artlbv The choice on closing this PR and making a new one is up to you. Whatever you find easiest. If you want to make a new PR go ahead and I'll give this a final look and sign. Otherwise, let me know and I'll give this PR a final look and sign here.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-46489/43401

cmsbuild commented:
Pull request #46489 was updated. @Moanwar, @aloeliger, @antoniovilela, @cmsbuild, @davidlange6, @epalencia, @fabiocos, @mandrenguyen, @rappoccio, @srimanob, @subirsarkar can you please check and sign again.

artlbv commented:
@cmsbuild please test

HaarigerHarald commented on L1Trigger/Configuration/python/Phase2GTMenus/SeedDefinitions/step1_2024_explicitSeeds/test/output_Phase2_L1T.root:
Were these files committed intentionally?

artlbv replied:
no.. that was by mistake. will remove! thanks for spotting

cmsbuild commented:
+1

**Size**: This PR adds an extra 1976KB to repository
**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-d2424a/43922/summary.html
**COMMIT**: 3afe32c672e6c63408a6270608870c5859901eaf
**CMSSW**: CMSSW_15_0_X_2025-01-22-1100/el8_amd64_gcc12
**Additional Tests**: HLT_P2_TIMING
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week1/cms-sw/cmssw/46489/43922/install.sh` to create a dev area with all the needed externals and cmssw changes.


* **HLT P2 Timing**: [chart](https://cmssdt.cern.ch/circles/web/piechart.php?data_name=hlt-p2-timing&resource=time_thread&filter=CMSSW_15_0_X_2025-01-22-1100&dataset=CMSSW_15_0_X_2025-01-22-1100-cmssw46489/el8_amd64_gcc12/76/Phase2Timing_resources)

## Comparison Summary



**Summary**:
 - You potentially [added 7 lines](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_15_0_X_2025-01-22-1100+d2424a/66576/validateJR/logRootQA-events.log) to the logs
 - [Reco comparison results: 128 differences found in the comparisons](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_15_0_X_2025-01-22-1100+d2424a/66576/validateJR.html)
 - DQMHistoTests: Total files compared: 49
 - DQMHistoTests: Total histograms compared: 3819085
 - DQMHistoTests: [Total failures: 41](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_15_0_X_2025-01-22-1100+d2424a/66576/validateJR.html)
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 3819024
 - DQMHistoTests: Total skipped: 20
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 48 files compared)
 - Checked 214 log files, 184 edm output root files, 49 DQM output files
 - [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_15_0_X_2025-01-22-1100+d2424a/66576/triggerResults): found differences in 4 / 47 workflows

artlbv committed:
Squashed commit of the following:

commit 3afe32c672e6c63408a6270608870c5859901eaf
Author: Artur Lobanov <Artur.Lobanov@cern.ch>
Date:   Fri Jan 17 16:18:01 2025 +0100

    Move P2GT to use the new 2024 menu by default

commit 39d4c3af79fbc92ab02e185497ca73ed17a4e631
Author: Artur Lobanov <Artur.Lobanov@cern.ch>
Date:   Fri Jan 17 16:10:03 2025 +0100

    Add missing file for P2gt 2024 menu with explicit seed definitions

commit 08c6b4b8ce47eb47b59313c4d92d747ae0d02ddd
Author: Artur Lobanov <Artur.Lobanov@cern.ch>
Date:   Fri Jan 17 14:30:54 2025 +0100

    Add refactored 2024 P2GT menu with object templates

commit c8654862a0cee57e743800ba2ee5fc457e7dce99
Author: Artur Lobanov <Artur.Lobanov@cern.ch>
Date:   Fri Jan 17 14:26:18 2025 +0100

    Add 2024 P2GT menu with object dictionaries to sync with Phase-2 L1 MenuTools

commit 97657455691fde7f925553ccc20df2adc649579c
Author: Artur Lobanov <Artur.Lobanov@cern.ch>
Date:   Fri Jan 17 14:22:18 2025 +0100

    Rename 2023 P2GT menu file to step1_2023

commit 651217d4d260a93d054928062f2a0f601161013e
Author: Artur Lobanov <Artur.Lobanov@cern.ch>
Date:   Fri Jan 17 14:11:30 2025 +0100

    Remove prototypeseeds as redundant

commit 9df945064904a3c3862907be0cc799a5c07eaba0
Author: Artur Lobanov <Artur.Lobanov@cern.ch>
Date:   Fri Jan 17 14:09:36 2025 +0100

    Move P2GT Menu into L1 Configurations

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-46489/43410

cmsbuild commented:
Pull request #46489 was updated. @Moanwar, @aloeliger, @antoniovilela, @cmsbuild, @davidlange6, @epalencia, @fabiocos, @mandrenguyen, @rappoccio, @srimanob, @subirsarkar can you please check and sign again.

artlbv commented:
@cmsbuild please test

artlbv commented:
@aloeliger this PR is ready now. I fixed the spurious file and squashed the commits.
I launched the tests for formal checks, though the content is the same as tested before.

please review whenever you can! 
thank you

cmsbuild commented:
+1

**Size**: This PR adds an extra 28KB to repository
**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-d2424a/43933/summary.html
**COMMIT**: 41a716c51305f326ad52186047a4de25b05a5f21
**CMSSW**: CMSSW_15_0_X_2025-01-23-1100/el8_amd64_gcc12
**Additional Tests**: HLT_P2_TIMING
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week1/cms-sw/cmssw/46489/43933/install.sh` to create a dev area with all the needed externals and cmssw changes.


* **HLT P2 Timing**: [chart](https://cmssdt.cern.ch/circles/web/piechart.php?data_name=hlt-p2-timing&resource=time_thread&filter=CMSSW_15_0_X_2025-01-23-1100&dataset=CMSSW_15_0_X_2025-01-23-1100-cmssw46489/el8_amd64_gcc12/77/Phase2Timing_resources)

## Comparison Summary



**Summary**:
 - You potentially [added 5 lines](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_15_0_X_2025-01-23-1100+d2424a/66585/validateJR/logRootQA-events.log) to the logs
 - [Reco comparison results: 120 differences found in the comparisons](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_15_0_X_2025-01-23-1100+d2424a/66585/validateJR.html)
 - DQMHistoTests: Total files compared: 49
 - DQMHistoTests: Total histograms compared: 3819085
 - DQMHistoTests: [Total failures: 103](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_15_0_X_2025-01-23-1100+d2424a/66585/validateJR.html)
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 3818962
 - DQMHistoTests: Total skipped: 20
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 48 files compared)
 - Checked 214 log files, 184 edm output root files, 49 DQM output files
 - [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_15_0_X_2025-01-23-1100+d2424a/66585/triggerResults): found differences in 4 / 47 workflows

aloeliger commented:
+l1

cmsbuild commented:
This pull request is fully signed and it will be integrated in one of the next master IBs (tests are also fine). This pull request will now be reviewed by the release team before it's merged. @sextonkennedy, @antoniovilela, @mandrenguyen, @rappoccio (and backports should be raised in the release meeting by the corresponding L2)

mandrenguyen commented:
+1
